expose/proxy_test.go

```diff
@@ -189,10 +189,11 @@ func TestProxyHTTPWebsocket(t *testing.T) {
 
 	url := fmt.Sprintf("ws://127.0.0.1:%d", listenPort)
 	dialer := websocket.Dialer{}
-	ws, _, err := dialer.Dial(url, nil)
+	ws, resp, err := dialer.Dial(url, nil)
 	if err != nil {
 		t.Fatalf("client Dial: %s", err)
 	}
+	defer resp.Body.Close()
```

Comment:
Maybe. The reason I preferred to defer is that imagine someone comes alone and needs to do something with `resp`. It would be quite easy for them to add code directly before the `resp.Body.Close()` line that did something with `resp`, and not notice any problems. But they will have introduced a bug, because they could easily not notice that they need to change `resp.Body.Close()` to `defer resp.Body.Close()`. By having the `defer` statement there from the beginning, there is no risk that this happens. What is the disadvantage of closing `resp` at the end of the function? I don't think it comes at much of an additional cost, but helps to future-proof the code a little.

I can change if you really want, but I think both versions have their merits.

Reply:
I think your point is reasonable. Reading from a closed reader is a panic, so in this hypothetical you'd notice, but I've come around.

Reply:
Just to be clear, I mean a slightly different scenario. Let's say the code looks initially like this, without the `defer` statement:

```
	ws, resp, err := dialer.Dial(url, nil)
	if err != nil {
		t.Fatalf("client Dial: %s", err)
	}
	resp.Body.Close()
```

A new requirement comes in, that makes Tuesdays special, such that no further work needs to be done on a Tuesday. Perhaps something like the following gets added:

```
	ws, resp, err := dialer.Dial(url, nil)
	if err != nil {
		t.Fatalf("client Dial: %s", err)
	}
	// Bug 1234567: tuesdays are special
	if resp.Header.get("Day-of-Week") == "Tuesday" {
		return
	}
	resp.Body.Close()
```

This change won't result in a runtime panic since the bug is slightly more subtle: on Tuesdays, the response body won't be closed. Of course, they may notice that they need to change the line below to be in a defer block, but it is easy to miss.

If it started life as this:

```
	ws, resp, err := dialer.Dial(url, nil)
	if err != nil {
		t.Fatalf("client Dial: %s", err)
	}
	defer resp.Body.Close()
```

then the change is more likely to be applied as:

```
	ws, resp, err := dialer.Dial(url, nil)
	if err != nil {
		t.Fatalf("client Dial: %s", err)
	}
	defer resp.Body.Close()
	if resp.Header.get("Day-of-Week") == "Tuesday" {
		return
	}
```

Of course it is still possible that the code gets added before the `defer` statement (which would also be wrong), so having the `defer` from the start isn't foolproof, but I think it is easier to spot when adding the new hypothetical feature.

It can also be that I am overthinking the problem! :-)